### PR TITLE
lxd/instance: Fix exec record-output location

### LINKED ISSUE #11643: Move exec record-output files to instance storage
Currently an exec session using `record-output=true` will write its stdout and stderr inside of LogPath, the issue is that LogPath isn't restricted by instance quota, allowing for a potential DoS of the host storage.

We should instead create a folder like `exec` or `exec-output` inside of the instance storage and record the exec output there instead.

### PR DESCRIPTION
**Pull request by MusicDin**
Store the output of an exec command (when `record-output` is true) in the instance storage instead of the log directory, as instance's storage quota is not applied to the log directory.

Fixes #11643

### CODE REVIEW

**Review by tomponline — CHANGES_REQUESTED** [review]
Please can you split this into 2 logical commits, one for the change in `lxd/instance...` and one for the use of the function in `lxd/instance_exec`. Thanks

**Comment by tomponline** [comment]
jenkins: test this please

**Comment by MusicDin** [comment]
> Please can you split this into 2 logical commits...

Absolutely

**Review by tomponline — APPROVED** [review]
LGTM, thanks!

**Review by tomponline — CHANGES_REQUESTED** [review]
Tests are failing:

https://jenkins.linuxcontainers.org/job/lxd-github-pull-test/9781/arch=amd64,backend=dir,compiler=golang-1.18,section=standalone/console#console-section-5

```
lxc query /1.0/instances/x1/logs/exec_8fd1a797-5b8d-4a88-803f-7a4bbcd81288.stdout --verbose
Error: open /lxc-ci/build/tmp.2H2ngLIFnY/go/src/github.com/lxc/lxd/test/tmp.TXE/ReI/logs/x1/exec_8fd1a797-5b8d-4a88-803f-7a4bbcd81288.stdout: no such file or director
```

**Comment by MusicDin** [comment]
The exec output is now stored on the instance's storage, however, the test checks for a file in the logs directory (`.../logs/x1/exec_*.stdout`).

Could it be that the returned paths in the metadata section have to be adjusted (https://github.com/lxc/lxd/pull/11791/files#diff-40362e9130fb8a2eb093355b5d3799af607edc65862b2834ed5060db36556241L701-L705)?

Currently:
```json
"output": {
  "1": "/1.0/instances/<instance>/logs/exec_*.stdout",
  "2": "/1.0/instances/<instance>/logs/exec_*.stderr"
},
```

**Comment by tomponline** [comment]
> The exec output is now stored on the instance's storage, however, the test checks for a file in the logs directory (`.../logs/x1/exec_*.stdout`).
> 
> Could it be that the returned paths in the metadata section have to be adjusted (https://github.com/lxc/lxd/pull/11791/files#diff-40362e9130fb8a2eb093355b5d3799af607edc65862b2834ed5060db36556241L701-L705)?
> 
> Currently:
> 
> ```json
> "output": {
>   "1": "/1.0/instances/<instance>/logs/exec_*.stdout",
>   "2": "/1.0/instances/<instance>/logs/exec_*.stderr"
> },
> ```

Mmm interesting. @stgraber how would you like us to proceed here? Shall we update the URLs returned here to be in the form:

```
/1.0/instances/<instance>/logs/exec-output/....
```

And then put an API endpoint/rule in place to load those files from the correct place?

**Comment by MusicDin** [comment]
@stgraber How should I extend the existing API endpoint rule?

Allow `exec-output` subpath:
```sh
Path: instances/{name}/logs/{file} -> instances/{name}/logs/{file:.*}
# Could be also more restrictive to allow only `exec-output/` subpath
```

Or create a new endpoint or follow some other approach that I have overlooked?

**Comment by tomponline** [comment]
> ```shell
> /logs/
> ```

I think I would prefer to see a separate endpoint, but pointing to the same function handlers. That way we won't get any issues with `./..` etc.

**Comment by tomponline** [comment]
@MusicDin while you have your hand in the basket, please could you also modify `instanceLogDelete` to change the check for:

```
	if file == "lxc.log" || file == "lxc.conf" {
		return response.BadRequest(fmt.Errorf("lxc.log and lxc.conf may not be deleted"))
	}
```

To check for `*.conf` and `qemu.log` and `lxc.log`. Thanks

**Comment by tomponline** [comment]
@MusicDin actually please can you change that `instanceLogDelete` function to only allow deletion of `*.log`, excluding `lxc.log` and `qemu.log`.

**Comment by MusicDin** [comment]
@tomponline Can `exec-output` records be deleted? 

If so, then we also need to allow `exec_*.stdout` and `exec_*.stderr` files to be deleted.

**Comment by tomponline** [comment]
> @tomponline Can `exec-output` records be deleted?
> 
> If so, then we also need to allow `exec_*.stdout` and `exec_*.stderr` files to be deleted.

Yes they do.

File: lxd/instance_logs.go [hunk 1]
```diff
@@ -322,15 +335,31 @@ func instanceLogDelete(d *Daemon, r *http.Request) response.Response {
...
+	if !validLogFileName(fileName) {
+		return response.BadRequest(fmt.Errorf("log file name %s not valid", fileName))
+	}
+
+	if !strings.HasSuffix(fileName, ".log") &&
+		!strings.HasSuffix(fileName, ".stdout") &&
+		!strings.HasSuffix(fileName, ".stderr") ||
+		fileName == "lxc.log" ||
+		fileName == "qemu.log" {
+		return response.BadRequest(fmt.Errorf("only log files excluding qemu.log and lxc.log may be deleted"))
```
Comment (tomponline) [inline lxd/instance_logs.go:349]: upper case first letter please.

File: lxd/instance_logs.go [hunk 2]
```diff
@@ -322,15 +335,31 @@ func instanceLogDelete(d *Daemon, r *http.Request) response.Response {
...
-	if file == "lxc.log" || file == "lxc.conf" {
-		return response.BadRequest(fmt.Errorf("lxc.log and lxc.conf may not be deleted"))
+	var filePath string
+	switch filepath.Dir(file) {
+	case ".":
+		filePath = shared.LogPath(project.Instance(projectName, name), fileName)
+	case "exec-output":
+		filePath = filepath.Join(inst.ExecOutputPath(), fileName)
+	default:
+		return response.BadRequest(fmt.Errorf("log path %s not valid", file))
```
Comment (tomponline) [inline lxd/instance_logs.go:359]: upper case first letter please.
And %q please

File: lxd/instance_logs.go [hunk 3]
```diff
@@ -238,13 +239,25 @@ func instanceLogGet(d *Daemon, r *http.Request) response.Response {
...
+	if !validLogFileName(fileName) {
+		return response.BadRequest(fmt.Errorf("log file name %s not valid", fileName))
+	}
+
+	var filePath string
+	switch filepath.Dir(file) {
+	case ".":
+		filePath = shared.LogPath(project.Instance(projectName, name), fileName)
+	case "exec-output":
+		filePath = filepath.Join(inst.ExecOutputPath(), fileName)
```
Comment (tomponline) [inline lxd/instance_logs.go:253]: Can you test what the old behaviour was, but I think it was possible to do an `lxc exec` then stop the instance and then be able to retrieve the output after stopping the instance. 

Now we are storing the exec output underneath the instance's root volume, in order for the files to be accessible even when the instance is off (for all storage pool types) we need to mount the instance's volume first in this case.

**Review by tomponline — CHANGES_REQUESTED** [review]
Thinking more about this @MusicDin I think we will need to support listing the exec output files in the listing endpoint still, as otherwise there would be no way to discover them in order to delete them.

File: lxd/instance_logs.go (hunk 3 above)
Comment (MusicDin) [reply lxd/instance_logs.go:253]: I have tested with storage pools of type `dir` and `zfs`. When an instance is stopped it works only for an instance with storage type of `dir`.

**Comment by tomponline** [comment]
@MusicDin see what you think but it feels like we may be ending up heading towards needing separate API handlers for `exec-output` endpoints, so that we can mount the instance and not make the other API handlers more complicated.

To mount the instance volume you can use `func LoadByInstance(s *state.State, inst instance.Instance)` and then `pool.MountInstance()` followed by `pool.UnmountInstance()`.

File: lxd/instance_logs.go (hunk 3 above)
Comment (tomponline) [reply lxd/instance_logs.go:253]: Is this now or before?

**Comment by MusicDin** [comment]
It seems feasible to return all available logs (including exec-output).
I will create new endpoints and update the PR once I'm done.

**Comment by MusicDin** [comment]
@tomponline Is it possible to stop pending tests since they are currently not required?

**Comment by tomponline** [comment]
> @tomponline Is it possible to stop pending tests since they are currently not required?

No im afraid not. @stgraber has been meaning to add cancellation support for years, but its dependent on the actual workers supporting being cancelled as I understand it.

File: lxd/instance_logs.go (hunk 3 above)
Comment (MusicDin) [reply lxd/instance_logs.go:253]: I've tested and I can confirm that in previous implementation it works perfectly fine when the container is stopped (both getting and deleting the exec-output)

**Comment by MusicDin** [comment]
@tomponline 

Here I mount the instance: https://github.com/lxc/lxd/compare/d175c61588e585895499a35cf14eee7557846cbf..24afffbb1d00f2b2585d1bfed12c47ab370d4fc9#diff-579c72b02f9858ac734c7a59fa6a3577b3aa66ddc8d7840b8b21e578cbb71f4aR436-R447

So in the logs it says that instance is mounted on the:
```
Mounted ZFS dataset dev=default/containers/c3 driver=zfs path=/var/snap/lxd/common/lxd/storage-pools/default/containers/c3
```

And `inst.ExecOutputPath()` returns the path:
```
/var/snap/lxd/common/lxd/containers/c3/exec-output
```

But `/var/snap/lxd/common/lxd/containers/c3` is effectively a link to the `/var/snap/lxd/common/lxd/storage-pools/default/containers/c3`.

Am I missing something?

**Comment by tomponline** [comment]
You're returning the FileResponse, which by the time it tries to read the file, the defer to unmount has already run.

**Comment by tomponline** [comment]
> FileResponse

I suggest updating `Render` to close the file before calling Cleanup.

File: lxd/response/response.go [hunk 4]
```diff
@@ -445,6 +445,11 @@ func (r *fileResponse) Render(w http.ResponseWriter) error {
 
 	for _, entry := range r.files {
 		var rd io.Reader
+
+		if entry.Cleanup != nil {
```
Comment (tomponline) [inline lxd/response/response.go:449]: @MusicDin I think we should leave this one alone, as in cases of multiple files I dont think its a good idea to stack up individual file cleanup functions as that changes the order of when things are run.

**Comment by tomponline** [comment]
@stgraber did the test suite run ok for this? It shows as pending, but the link goes back to github weirdly.

**Comment by MusicDin** [comment]
@tomponline We changed the commit message of the last commit, this is why the tests were run again. 
Jenkins results before changing the commit message are here: https://jenkins.linuxcontainers.org/job/lxd-github-pull-test/9814/

**Comment by tomponline** [comment]
Very good!

**Pull request merged**